scipy/sparse/linalg/_isolve/iterative.py

```diff
@@ -670,14 +671,14 @@ def gmres(A, b, x0=None, tol=None, restart=None, maxiter=None, M=None,
 
     # Handle the deprecation frenzy
     if restrt and restart:
```

Comment:
Test failures are a result of a bug here. `_NoValue` evaluates trues so we always incorrectly enter this block when `restart` is provided.
```suggestion
    if restrt not in (None, _NoValue) and restart:
```

Reply:
Nice. Thanks for catching that.